### PR TITLE
Flink Batch Runner: optimize GBK translation for non-merging windows.

### PR DESCRIPTION
**Pull request by dmvk**
https://issues.apache.org/jira/browse/BEAM-8848

dependes on https://github.com/apache/beam/pull/11857

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/)

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by dmvk** [comment]
Run Flink ValidatesRunner

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java [hunk 1]
```diff
@@ -105,40 +111,53 @@
   // --------------------------------------------------------------------------------------------
 
   @SuppressWarnings("rawtypes")
-  private static final Map<String, FlinkBatchPipelineTranslator.BatchTransformTranslator>
+  private static final Map<String, List<FlinkBatchPipelineTranslator.BatchTransformTranslator>>
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java:114]: We might want to use `Multimap`.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java [hunk 2]
```diff
@@ -105,40 +111,53 @@
...
       TRANSLATORS = new HashMap<>();
 
-  static {
-    TRANSLATORS.put(PTransformTranslation.IMPULSE_TRANSFORM_URN, new ImpulseTranslatorBatch());
+  private static void registerTranslator(
+      String urn, FlinkBatchPipelineTranslator.BatchTransformTranslator<?> translator) {
+    if (!TRANSLATORS.containsKey(urn)) {
+      TRANSLATORS.put(urn, new ArrayList<>());
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java:120]: Again, multimap would handle this.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java [hunk 3]
```diff
@@ -105,40 +111,53 @@
...
+    }
+    TRANSLATORS.get(urn).add(translator);
+  }
 
-    TRANSLATORS.put(
+  static {
+    registerTranslator(PTransformTranslation.IMPULSE_TRANSFORM_URN, new ImpulseTranslatorBatch());
+    registerTranslator(
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java:127]: I think we can get rid of these helper calls and keep the existing code.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTranslationContext.java [hunk 4]
```diff
@@ -68,13 +71,18 @@ public FlinkBatchTranslationContext(ExecutionEnvironment env, PipelineOptions op
...
 
+  void init(Pipeline pipeline) {
+    pipeline.traverseTopologically(countingPipelineVisitor);
+    pipeline.traverseTopologically(lookupPipelineVisitor);
+  }
+
   // ------------------------------------------------------------------------
 
-  public Map<PValue, DataSet<?>> getDanglingDataSets() {
+  Map<PValue, DataSet<?>> getDanglingDataSets() {
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTranslationContext.java:81]: Can you revert these? This makes integration hard for external Runners.

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkNonMergingReduceFunction.java [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+      this.iterator = iterator;
+    }
+
+    @Override
+    public Iterator<T> iterator() {
+      if (used.compareAndSet(false, true)) {
+        return iterator;
+      }
+      throw new IllegalStateException(
+          "GBK result is not re-iterable. You can enable re-iterations by setting '--reIterableGroupByKeyResult=true'.");
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkNonMergingReduceFunction.java:64]:
```suggestion
          "GBK result is not re-iterable. You can enable re-iterations by setting '--reIterableGroupByKeyResult'.");
```

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java (hunk 1 above)
Comment (dmvk) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java:114]: 👍

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTranslationContext.java (hunk 4 above)
Comment (dmvk) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTranslationContext.java:81]: The whole class is marked as package private, so this change should not make any difference when it comes to 3rd code. Am I missing something?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkNonMergingReduceFunction.java (hunk 5 above)
Comment (dmvk) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/translation/functions/FlinkNonMergingReduceFunction.java]: 👍

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java [hunk 6]
```diff
@@ -105,40 +113,45 @@
...
-        PTransformTranslation.FLATTEN_TRANSFORM_URN, new FlattenPCollectionTranslatorBatch());
-
+        PTransformTranslation.GROUP_BY_KEY_TRANSFORM_URN, new GroupByKeyTranslatorBatch<>());
+    TRANSLATORS.put(PTransformTranslation.RESHUFFLE_URN, new ReshuffleTranslatorBatch<>());
     TRANSLATORS.put(
-        PTransformTranslation.ASSIGN_WINDOWS_TRANSFORM_URN, new WindowAssignTranslatorBatch());
-
-    TRANSLATORS.put(PTransformTranslation.PAR_DO_TRANSFORM_URN, new ParDoTranslatorBatch());
-
-    TRANSLATORS.put(PTransformTranslation.READ_TRANSFORM_URN, new ReadSourceTranslatorBatch());
```
Comment (mxm) [inline runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTransformTranslators.java:132]: Could we keep the original order of the translators here?

File: runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTranslationContext.java (hunk 4 above)
Comment (mxm) [reply runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkBatchTranslationContext.java:81]: Didn't see that this was the case. It's still an unrelated change and it draws attention away from the actual issue.

**Comment by dmvk** [comment]
Run Flink ValidatesRunner

**Pull request merged**